udp/client/clientconn.go

```diff
@@ -614,8 +643,8 @@ func (cc *ClientConn) getResponseFromCache(mid int32, resp *pool.Message) (bool,
 // CheckMyMessageID compare client msgID against peer messageID and if it is near < 0xffff/4 then incrase msgID.
 // When msgIDs met it can cause issue because cache can send message to which doesn't bellows to request.
 func (cc *ClientConn) CheckMyMessageID(req *pool.Message) {
-	if req.Type() == message.Confirmable && uint16(req.MessageID())-uint16(atomic.LoadUint32(&cc.msgID)) < 0xffff/4 {
-		atomic.AddUint32(&cc.msgID, 0xffff/2)
+	if req.Type() == message.Confirmable && uint16(req.MessageID())-uint16(cc.msgID.Load()) < 0xffff/4 {
+		cc.msgID.Add(0xffff / 2)
```

Comment:
Time of check vs time of use race:

ThreadA executes line 646 and get rescheduled, ThreadB does the same and both increase the msgID by `0xffff / 2`. Don't know if that's a problem, if it is we need to use CAS here:

```go
if req.Type() == message.Confirmable {
  for {  
    oldID := cc.msgID.Load()
    if uint16(req.MessageID())-uint16(cc.msgID.Load()) >= 0xffff/4 {
      return
    }
    newID := oldID + 0xffff / 2
    if cc.msgID.CAS(oldID, newID) {
      break
    }
  }

}

```


Also msgID can be -1 or some int32_t value that's invalid for int16_t, right? Shouldn't that be checked for?

Reply:
`msgID` is uint32 ab it is always cast to uint16